Src/ILGPU/IR/Transformations/CodePlacement.cs

```diff
@@ -0,0 +1,876 @@
+            public void SetupInsertPosition(BasicBlock.Builder builder) =>
+                builder.SetupInsertPositionToStart();
+        }
+
+        #endregion
+
+        #region Movement and Placement
+
+        /// <summary>
+        /// Tracks and validates the movement of values with side effects to different
+        /// blocks. Enabling movement of side effect values makes the whole placement
+        /// algorithm significantly more aggressive.
+        /// </summary>
+        private readonly struct Mover : IMover
+        {
+            #region Instance
+
+            /// <summary>
+            /// Maps side effect values to their original source blocks in which they
+            /// have been defined
+            /// </summary>
+            private readonly Dictionary<SideEffectValue, BasicBlock> valueBlocks;
+
+            /// <summary>
+            /// Stores all memory values according to the reverse post order.
+            /// </summary>
+            private readonly List<MemoryValue> values;
+
+            /// <summary>
+            /// Maps all memory values to their global indices in the <see cref="values"/>
+            /// list determined by a reverse-post-order search.
+            /// </summary>
+            private readonly Dictionary<MemoryValue, int> valueIndices;
+
+            /// <summary>
+            /// Stores all end indices of all blocks pointing to offsets in the
+            /// <see cref="values"/> list.
+            /// </summary>
+            private readonly BasicBlockMap<int> blockRanges;
+
+            /// <summary>
+            /// Constructs a new mover.
+            /// </summary>
+            /// <param name="builder">The parent builder.</param>
+            /// <param name="mode">The current placement mode.</param>
+            [SuppressMessage(
+                "Maintainability",
+                "CA1508:Avoid dead conditional code",
+                Justification = "There is no dead code in the method below")]
+            public Mover(Method.Builder builder, CodePlacementMode mode)
+            {
+                // Setup internal mappings to store value indices and block offsets
+                var blocks = builder.SourceBlocks;
+                values = new List<MemoryValue>(blocks.Count << 1);
+                valueBlocks = new Dictionary<SideEffectValue, BasicBlock>(
+                    values.Capacity);
+                valueIndices = new Dictionary<MemoryValue, int>(values.Capacity);
+                blockRanges = blocks.CreateMap<int>();
+
+                // Gather all side effect values in the order of their appearance
+                foreach (var block in blocks)
+                {
+                    foreach (Value value in block)
+                    {
+                        if (!(value is SideEffectValue sev))
+                            continue;
+
+                        valueBlocks[sev] = value.BasicBlock;
+                        if (value is MemoryValue memoryValue)
```

Comment:
QUESTION: Is this MemoryValue correct? Should it be SideEffectValue?
```suggestion
                        if (value is SideEffectValue sideEffectValue)
```

Reply:
This should be correct as `MemoryValues` have to appear in a pre-defined order (in order to preserve read-write dependencies). Other values with side effects can depend on the control flow itself (e.g. a `WarpShuffle` value). However, they can be freely moved around as long as they do not end up in a block that violates their "current" control-flow preconditions.